install.sh

```diff
@@ -59,29 +62,52 @@ fi
 
 # Downloading artifact
 mkdir -p "${BIN_FOLDER}"
-curl "${RELEASE_URL}" -o "${BIN_FOLDER}/micromamba" -fsSL --compressed ${CURL_OPTS:-}
+if hash curl >/dev/null 2>&1; then
+  curl "${RELEASE_URL}" -o "${BIN_FOLDER}/micromamba" -fsSL --compressed ${CURL_OPTS:-}
+elif hash wget >/dev/null 2>&1; then
+  wget ${WGET_OPTS:-} -qO "${BIN_FOLDER}/micromamba" "${RELEASE_URL}"
+else
+  echo "Neither curl nor wget was found" >&2
+  exit 1
+fi
 chmod +x "${BIN_FOLDER}/micromamba"
 
 
 # Initializing shell
-if [[ "$INIT_YES" == "" || "$INIT_YES" == "y" || "$INIT_YES" == "Y" || "$INIT_YES" == "yes" ]]; then
-  case "$("${BIN_FOLDER}/micromamba" --version)" in
-    1.*|0.*)
-      "${BIN_FOLDER}/micromamba" shell init -p "${PREFIXLOCATION}"
-      ;;
-    *)
-      "${BIN_FOLDER}/micromamba" shell init --root-prefix "${PREFIXLOCATION}"
-      ;;
-  esac
-
-  echo "Please restart your shell to activate micromamba or run the following:\n"
-  echo "  source ~/.bashrc (or ~/.zshrc, ...)"
-fi
+case "$INIT_YES" in
+  y|Y|yes)
+    case "`"${BIN_FOLDER}/micromamba" --version`" in
+      1.*|0.*)
+        shell_arg=-s
+        prefix_arg=-p
+        ;;
+      *)
+        shell_arg=--shell
+        prefix_arg=--root-prefix
+        ;;
+    esac
+
+    SHELLS=
+    [ -e $HOME/.bashrc ] && SHELLS="$SHELLS bash"
+    [ -e "${XDG_CONFIG_HOME:-$HOME/.config}/fish" ] && SHELLS="$SHELLS fish"
+    [ -e "$HOME/.xonshrc" -o -e "${XDG_CONFIG_HOME:-$HOME/.config}/xonsh" ] && SHELLS="$SHELLS xonsh"
+    [ -e "${ZDOTDIR:-$HOME}/.zshrc" ] && SHELLS="$SHELLS zsh"
+
+    for shell in $SHELLS; do
+        "${BIN_FOLDER}/micromamba" shell init $shell_arg $shell $prefix_arg "$PREFIX_LOCATION"
+    done
```

Comment:
I'm not sure if initializing all shells is better than only the current one.
@jonashaag ?

Reply:
No, definitely not initialize all shells, this will definitely cause lots of confusion and trouble for people less knowledgeable about shells

Reply:
The problem is that `fish` and `xonsh` are not POSIX compliant, so you cannot run this script with `fish install.sh` or `xonsh install.sh`. If you are using one of these shells, you will still need to run `install.sh` using (explicitly or implicitly) `sh` or `bash`. But in this case, the script doesn't know which interactive shell the user is actually using. I don't think there's a reliable, non-hacky way to detect it, and anyway, the user can switch between shells.

I, for one, switch between `bash`, `zsh` and `xonsh` depending on the circumstances, and I want `micromamba` to work in all of them, so I have it initialised in the startup files for all these shells. What kind of confusion and problems can this cause? I haven't noticed any.

I think people who aren't knowledgeable about shells just stick to the default one, so there's no harm in initialising `micromamba` for a shell they don't use. And anyway, if a less knowledgeable user has always used `bash`, and one day tries to run `zsh` or `xonsh`, they might be confused as to why `micromamba activate`, etc. works in `bash`, but doesn't in `zsh`/`xonsh`.

If all available shells are initialised, the installation instructions can be exactly the same for everyone: `sh install.sh` (or `./install.sh`, if we already know that `install.sh` is executable). Otherwise, there is no one single installation command that works in `fish`, `xonsh` and `bash`, and only initialises the shell it was called from.

I'd be happy to improve `install.sh` further to address your concerns, but I really think the solution should be equally transparent for all shells.

Reply:
So the question is, if my login shell (`$SHELL`) is `zsh`, but I run `install.sh` from `bash`, which shell should be activated? (It's not a made up example, I have a similar situation at work, I don't have permission to change the shell from `bash` to `xonsh` at work, so I have to run `xonsh` manually or in a startup script.)

So far, the obvious answer for me was that if one shell should be initialised, it should be the shell that `install.sh` was called from, since this is clearly a shell that the user actually uses. It looks like the answer for you is that it should be the login shell. Since different people have different intuitions, perhaps at least these two shells (the login shell and the calling shell) should be initialised? Or at least the script can ask which shells can be initalised?

If the installation script is called from the command line, the calling shell can actually be detected from the script with `ps -o comm $PPID |tail -1`.

How about this: if the above command returns a name of a supported shell, initialise both that shell and the login shell (`$SHELL`). Otherwise, just initialise the login shell. I understand that you want simplicity, but perhaps such a rule would be simple enough? I think it's intuitive: if the user is running `install.sh` from a shell, it's safe to assume that they want that shell to be initalised, even if it's not their login shell.

Reply:
Not the login shell per say. If I am using xonsh and I run `bash <(curl -L micro.mamba.pm/install.sh)`, I would expect it to initializae xonsh.

Reply:
OK, if so, we could just initialise the shell that was used to run the script. If we cannot detect it, we can fall back to `$SHELL`. What do you think?

```suggestion
    parent=$(ps -o comm $PPID |tail -1)
    parent="${parent#-}"
    case "$parent" in
      # shells supported by `micromamba shell init`:
      bash|fish|xonsh|zsh)
        shell=$parent
        ;;
      *)
        shell=${SHELL##*/}
        ;;
    esac
    "${BIN_FOLDER}/micromamba" shell init $shell_arg "$shell" $prefix_arg "$PREFIX_LOCATION"
```

Reply:
Could be something like this. What is the `#` and `##` expansion?
We would also need quotes around `parent` and `SHELL`  to be safe

Reply:
Shell initialization might make some other programs unusable if they run subprocesses through the shell or run shell scripts or what have you. Thus it's really important that people are aware we are changing something in their shells, and what shells we are changing.

How about we try to autodetect the shell(s) used by the user, present a list, and have the user confirm/edit?

```
Enter a the list of shells to initialize, eg: "bash zsh".
Enter an empty string to initialize no shells.
Default (autodetected): "bash fish"
```

Reply:
@AntoinePrv If the parent process is a login shell, it will have a `-` in front of its name in the `ps` output, e.g. `-zsh`. The `#` expansion removes the leading dash, if present. `$SHELL` contains the full path of the login shell, and we just want its name, so we remove all characters up to the last `/` (double `#` makes the `*` greedy).

Variable assignments in the `var1=$var2` format [don't need quotes](https://unix.stackexchange.com/a/719808/360988), both according to the POSIX standard and the actual `sh` implementations. I think omitting the quotes in places where they are not needed makes the script a bit more readable (we need to have a lot of quotes anyway).

Reply:
The advantage of the shell expansion over `tr` and `basename` is that it doesn't spawn new processes, but that's of course a minor thing these days.

OK, to sum up the discussion:
1. Shell detection: we could detect shells by looking at rc files present in the user's home directory (my original code), detect only the calling shell (my modified code), or detect both the calling shell and the login shell (`$SHELL`).
2. Confirmation: we can list the detected shells and allow the user to change this list (@jonashaag's suggestion), we can just silently initialise the shells (as in the code above), or we can we can add a list of detected shells to the "Init shell" question (e.g. `Init shells (bash, zsh)? [Y/n]`) and perhaps tell the user that they can run `micromamba shell init ...` for their shells manually if they answer `no`.

Please let me know which options you prefer, they all them seem fine. I think perhaps listing the detected shells would be preferable, as this lets the user know which shells are going to be initialised, but doesn't complicate the code with reading a shell list from the user.

Reply:
My preference is init only the calling shell.

Reply:
OK, done!